### PR TITLE
Migrate ProfilingLayer to tracing registry API

### LINKED ISSUE #8906: Migrate CurrentSpan to use the registry API
[`CurrentSpan`](https://docs.rs/tracing-subscriber/0.2.18/tracing_subscriber/struct.CurrentSpan.html) is being deprecated as of version `0.2.18` of `tracing-subscriber`, and according to the [release notes](https://github.com/tokio-rs/tracing/pull/1384), the substitute for it is the registry API.

Files: `client/tracing/src/lib.rs`, `client/tracing/src/block/mod.rs`

### PR DESCRIPTION
**Pull request by Folyd**
The `tracing` crate deprecated the `CurrentSpan` and will remove it in the future version. 
This PR migrates the ProfilingLayer to tracing [registry](https://docs.rs/tracing-subscriber/0.2.18/tracing_subscriber/registry/index.html) API.

polkadot companion: https://github.com/paritytech/polkadot/pull/3203
Fixes #8906.

### CODE REVIEW

**Comment by dvdplm** [comment]
cc @insipx @emostov

**Review by bkchr — APPROVED** [review]

**Review by insipx — APPROVED** [review]

**Comment by insipx** [comment]
LGTM

**Comment by Folyd** [comment]
> LGTM. I think `current_span` is also used in `tracing/src/block/mod.rs` (in the Subscriber used for the `trace_block` RPC). I'm guessing this will be a separate PR. Not sure if there is a function like `lookup_current_span()` for the `Subscriber`, which might make it trickier to migrate?

I don't think the `current_span` field in the `BlockSubscriber` is useful. I believe we can remove those codes related to `current_span`. Also, we needn't migrate the `BlockSubscriber` to the `registry` API.

File: client/tracing/src/block/mod.rs [hunk 1]
```diff
@@ -142,16 +138,10 @@ impl Subscriber for BlockSubscriber {
...
-		// Not currently used
-		unimplemented!("record_follows_from is not implemented");
-	}
-
 	fn event(&self, event: &tracing::Event<'_>) {
 		let mut values = crate::Values::default();
 		event.record(&mut values);
-		let parent_id = event.parent().cloned()
-			.or_else(|| self.current_span.id());
+		let parent_id = event.parent().cloned();
```
Comment (emostov) [inline client/tracing/src/block/mod.rs:144]: I assume this is `None` if no parent?

**Comment by emostov** [comment]
I need to test this out against our the balance reconciliation logic to make sure none of our assumptions break.

Does anyone know if span Id should still correspond to the order the spans where entered?

**Comment by insipx** [comment]
> I need to test this out against our the balance reconciliation logic to make sure none of our assumptions break.
> 
> Does anyone know if span Id should still correspond to the order the spans where entered?

AFAIU from talking in their discord, the only thing that has changed is the current_span is apart of registry. It was apart of it before too, but they just didn't remove `CurrentSpan`, which is what the deprecation is all about. So the ID's should still correspond to the order in which they were entered

**Comment by bkchr** [comment]
> I need to test this out against our the balance reconciliation logic to make sure none of our assumptions break.
> 
> Does anyone know if span Id should still correspond to the order the spans where entered?

It would be nice to have them as a test here in Substrate :P

**Comment by emostov** [comment]
> It would be nice to have them as a test here in Substrate :P

Agreed - Is there a way a to mock a block that can then be executed in a test with tracing enabled?

**Comment by bkchr** [comment]
> > It would be nice to have them as a test here in Substrate :P
> 
> Agreed - Is there a way a to mock a block that can then be executed in a test with tracing enabled?

Best to check the `test-runtime` and their usage in tests.

File: client/tracing/src/block/mod.rs (hunk 1 above)
Comment (Folyd) [reply client/tracing/src/block/mod.rs:149]: Yes, it's ok for the root span to miss the `parent_id`?

**Comment by bkchr** [comment]
@Folyd tests are still not happy. And can you please stop using force pushes?

File: client/tracing/src/block/mod.rs (hunk 1 above)
Comment (emostov) [reply client/tracing/src/block/mod.rs:149]: Yes

**Comment by Folyd** [comment]
> @Folyd tests are still not happy. And can you please stop using force pushes?

Thanks for reminding me. There still a dependency conflict. Any idea?

**Comment by bkchr** [comment]
> > @Folyd tests are still not happy. And can you please stop using force pushes?
> 
> Thanks for reminding me. There still a dependency conflict. Any idea?

You would need a so called companion pr. Could you read the contribution doc on how it works?

**Comment by Folyd** [comment]
Hi @bkchr. I filed a companion PR and updated the description. What should I do next?

**Comment by bkchr** [comment]
@Folyd ty :)

**Pull request merged**